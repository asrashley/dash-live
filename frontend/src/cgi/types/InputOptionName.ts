export interface InputOptionName {
    shortName: string;
    fullName: string;
    cgiName: string;
}