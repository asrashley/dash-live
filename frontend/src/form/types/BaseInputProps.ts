import type { ReadonlySignal } from "@preact/signals";
import { StaticInputProps } from "./StaticInputProps";

export type BaseInputProps = {
  name: string;
  id: string;
  type: StaticInputProps["type"];
  className: string | ReadonlySignal<string>;
  title: string;
  placeholder?: string;
  "aria-describedby": string;
  disabled: ReadonlySignal<boolean>;
  onInput: (ev: Event) => void;
};

