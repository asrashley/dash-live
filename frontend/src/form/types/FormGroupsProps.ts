import { InputFormGroup } from "./InputFormGroup";
import { FormRowMode } from "./FormRowMode";
import { SetValueFunc } from "./SetValueFunc";
import { InputProps } from "./InputProps";

export interface FormGroupsProps extends Pick<InputProps, "data" | "disabledFields"> {
  groups: InputFormGroup[];
  expand?: string;
  mode: FormRowMode;
  layout?: number[];
  setValue: SetValueFunc;
}
