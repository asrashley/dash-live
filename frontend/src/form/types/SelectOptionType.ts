export interface SelectOptionType {
    name?: string;
    title: string;
    value: string;
}
