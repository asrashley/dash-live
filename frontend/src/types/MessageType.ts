export type MessageLevel = 'success' | 'danger' | 'warning' | 'info';

export interface MessageType {
    id: number;
    text: string;
    footer?: string;
    level: MessageLevel;
}
