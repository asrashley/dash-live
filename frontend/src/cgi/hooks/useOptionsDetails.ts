import { type ReadonlySignal, useSignal } from "@preact/signals";

import { fieldGroups } from "@dashlive/options";

import { InputOptionName } from "../types/InputOptionName";

function createOptionNames(): InputOptionName[] {
    const names: InputOptionName[] = [];
    fieldGroups.forEach(grp => {
        grp.fields.forEach(({ name, fullName, shortName }) => {
            names.push({
                cgiName: name,
                fullName,
                shortName,
            });
        });
    });
    return names;
}

export interface UseOptionsDetailsHook {
    allOptions: ReadonlySignal<InputOptionName[]>;
}

export function useOptionsDetails(): UseOptionsDetailsHook {
    const allOptions = useSignal<InputOptionName[]>(createOptionNames());
    return { allOptions };
}