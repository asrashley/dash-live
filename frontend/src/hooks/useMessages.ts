import { type Signal, signal, useComputed } from "@preact/signals";
import { MessageLevel, MessageType } from "../types/MessageType";

type MessagesState = {
  alerts: MessageType[];
  nextId: number;
}

const messages = signal<MessagesState>({ alerts: [], nextId: 1 });

// only exported for use in tests
export function resetAllMessages() {
  messages.value = { alerts: [], nextId: 1 };
}

export type AppendMessageFn = (text: string, level?: MessageLevel) => void;

const appendMessage: AppendMessageFn = (text: string, level: MessageLevel = "warning") => {
  const id = messages.value.nextId;
  const alerts = [
    ...messages.value.alerts,
    {
      text,
      id,
      level,
    },
  ];
  messages.value = {
    nextId: id + 1,
    alerts,
  };
};

const removeAlert = (id: number) => {
  const alerts = messages.value.alerts.filter((a) => a.id !== id);
  messages.value = { ...messages.value, alerts };
};

export interface UseMessagesHook {
  alerts: Signal<MessageType[]>
  appendMessage: (level: MessageLevel, text: string, footer?: string) => void;
  removeAlert: (id: number) => void;
}

export function useMessages(): UseMessagesHook {
  const alerts = useComputed(() => {
    return messages.value.alerts;
  });

  return { alerts, appendMessage, removeAlert };
}
