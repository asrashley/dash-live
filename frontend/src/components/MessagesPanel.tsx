import { useCallback } from "preact/hooks";

import { Alert } from "./Alert";
import { useMessages } from "../hooks/useMessages";

export function MessagesPanel() {
  const { alerts, removeAlert } = useMessages();

  const onDismiss = useCallback(
    (id: number) => {
      removeAlert(id);
    },
    [removeAlert]
  );

  return (
    <div className="messages-panel">
      {alerts.value.map((item) => (
        <Alert key={item.id} onDismiss={onDismiss} {...item} />
      ))}
    </div>
  );
}
