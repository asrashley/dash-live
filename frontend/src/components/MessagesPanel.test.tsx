import { afterEach, beforeEach, describe, expect, test, vi } from "vitest";
import { signal } from "@preact/signals";
import { fireEvent } from "@testing-library/preact";

import { renderWithProviders } from "../test/renderWithProviders";
import { MessagesPanel } from "./MessagesPanel";
import { useMessages } from "../hooks/useMessages";
import { MessageType } from "../types/MessageType";

vi.mock("../hooks/useMessages");

describe("MessagesPanel", () => {
  const useMessagesMock = vi.mocked(useMessages);
  const appendMessage = vi.fn();
  const removeAlert = vi.fn();
  const alerts = signal<MessageType[]>([]);

  beforeEach(() => {
    useMessagesMock.mockReturnValue({ alerts, appendMessage, removeAlert });
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  test("should display messages", () => {
    alerts.value = [
      {
        text: "alert message one",
        id: 12,
        level: "warning",
      },
      {
        text: "message two",
        id: 23,
        level: "info",
      },
    ];
    const { asFragment, getBySelector, getByText } = renderWithProviders(
      <MessagesPanel />
    );
    getBySelector(".messages-panel");
    getByText(alerts.value[0].text);
    getByText(alerts.value[1].text);
    expect(asFragment()).toMatchSnapshot();
  });

  test("can dismiss message", () => {
    alerts.value = [
      {
        text: "alert message one",
        id: 12,
        level: "warning",
      },
    ];
    const { getBySelector } = renderWithProviders(<MessagesPanel />);

    const btn = getBySelector("#alert_12 .btn-close");
    fireEvent.click(btn);
    expect(removeAlert).toHaveBeenCalledTimes(1);
    expect(removeAlert).toHaveBeenCalledWith(12);
  });
});
