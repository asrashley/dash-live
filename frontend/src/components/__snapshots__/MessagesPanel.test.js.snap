// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MessagesPanel > should display messages 1`] = `
<DocumentFragment>
  <div
    class="messages-panel"
  >
    <div
      class="alert alert-warning alert-dismissible fade show"
      id="alert_12"
      role="alert"
    >
      alert message one
      <button
        aria-label="Close"
        class="btn-close"
        data-bs-dismiss="alert"
        type="button"
      />
    </div>
    <div
      class="alert alert-info alert-dismissible fade show"
      id="alert_23"
      role="alert"
    >
      message two
      <button
        aria-label="Close"
        class="btn-close"
        data-bs-dismiss="alert"
        type="button"
      />
    </div>
  </div>
</DocumentFragment>
`;
